crates/bevy_utils/src/label.rs

```diff
@@ -60,90 +66,40 @@ where
     }
 }
 
-/// Macro to define a new label trait
-///
-/// # Example
+/// The [`TypeId`](std::any::TypeId) of a value and its [`DynHash`] output
+/// when hashed with the [`DefaultHasher`].
 ///
-/// ```
-/// # use bevy_utils::define_boxed_label;
-/// define_boxed_label!(MyNewLabelTrait);
-/// ```
-#[macro_export]
-macro_rules! define_boxed_label {
-    ($label_trait_name:ident) => {
-        /// A strongly-typed label.
-        pub trait $label_trait_name: 'static + Send + Sync + ::std::fmt::Debug {
-            /// Return's the [TypeId] of this label, or the the ID of the
-            /// wrappped label type for `Box<dyn
-            #[doc = stringify!($label_trait_name)]
-            /// >`
-            ///
-            /// [TypeId]: std::any::TypeId
-            fn inner_type_id(&self) -> ::std::any::TypeId {
-                std::any::TypeId::of::<Self>()
-            }
-
-            /// Clones this `
-            #[doc = stringify!($label_trait_name)]
-            /// `
-            fn dyn_clone(&self) -> Box<dyn $label_trait_name>;
-
-            /// Casts this value to a form where it can be compared with other type-erased values.
-            fn as_dyn_eq(&self) -> &dyn ::bevy_utils::label::DynEq;
-
-            /// Feeds this value into the given [`Hasher`].
-            ///
-            /// [`Hasher`]: std::hash::Hasher
-            fn dyn_hash(&self, state: &mut dyn ::std::hash::Hasher);
-        }
+/// This is expected to have sufficient entropy to be different for each value.
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct UniqueValue(TypeId, u64);
```

Comment:
My #1 concern with this approach is whether or not we can rely on this being a truly unique value. What would the error case look like if an unlucky user saw two of their labels. Can we estimate the probability of this occurring?

Reply:
In theory, yes we can estimate it. I don't know the exact math though. It depends on the hash functions used.

If we can assume all hashes are equally likely, then we can [approximate][1] the number of different things `n` we'd need to hash to have `p` probability of a collision (where `p <= 0.5`).

```
n = sqrt(2 * 2^(bits in hash) * p)
```

So if a hash is 64 bits and all hashes are equally likely, you'd need to hash `10^37` things to have a 1% chance of collision. 

So it's extremely unlikely. 

Rust actually extended `TypeId` to 128-bits recently to shrink the already astronomically small odds of collisions because they did encounter a couple. Rust programs have no choice but to consider a `TypeId` collision impossible, so we can only worry about value hash collisions, but worrying about it isn't justified by a mathematical argument.

[1]: https://en.wikipedia.org/wiki/Birthday_attack

Reply:
#7762 might be better about this, idk.